### PR TITLE
Add types to `T::Struct.new`

### PR DESCRIPTION
**Pull request by azdavis**
This makes subclasses of `T::Struct` have good static types for `new`.

I would recommend reviewing by commit.

### Motivation

To increase typedness.

### Test plan

See included automated tests.

### CODE REVIEW

File: dsl/ChalkODMProp.cc [hunk 1]
```diff
@@ -169,21 +169,27 @@ vector<unique_ptr<ast::Expression>> ChalkODMProp::replaceDSL(core::MutableContex
...
+        if (!hasForeign) {
             auto [fk, foreignTree] = ASTUtil::extractHashValue(ctx, *rules, core::Names::foreign());
             foreign = move(foreignTree);
             if (auto body = thunkBody(ctx, foreign.get())) {
                 foreign = std::move(body);
+                hasForeign = true;
             }
         }
+        auto hasDefault = ASTUtil::hasHashValue(ctx, *rules, core::Names::default_());
+        ret.prop.optional = hasDefault || hasForeign || hasComputedBy;
```
Comment (azdavis) [inline dsl/ChalkODMProp.cc]: After a reading of https://github.com/sorbet/sorbet/blob/master/gems/sorbet-runtime/lib/types/props/_props.rb#L31, I think this logic for determining whether an arg to `new` is optional or not is not correct. I think it might be:

```
arg is optional iff
rules[:factory] is present OR
rules[:default] is present OR
rules[:optional] is not false
```

where `rules` is all the stuff after the first argument passed to `prop` (or `const`).

**Comment by azdavis** [comment]
Oh, I just remembered I need to check whether the superclass in the ClassDef is `T::Struct`. If not, we don't want to synthesize an `initialize`.

**Comment by DarkDimius** [comment]
This seems to refactor a lot of code additionally to actual changes in behavior.
Can we land prework of refactors(that are not supposed to change behavior) as a separate PR?

File: dsl/ChalkODMProp.cc (hunk 1 above)
Comment (azdavis) [reply dsl/ChalkODMProp.cc]: Hm. Logic for `computed_by` is in ChalkODMProp.cc but the `computed_by` prop is not mentioned in that doc comment.

**Comment by azdavis** [comment]
Sure. What were you thinking could be landed as pre-work? Maybe everything except the logic that actually creates the synthesized initialize method (i.e., everything except the block inside `synthesizeInitialize`) and the tests?

File: dsl/ChalkODMProp.cc (hunk 1 above)
Comment (azdavis) [reply dsl/ChalkODMProp.cc]: After IRL with @jez, there's another condition for when the arg is optional: if the type of the arg is nilable.

**Comment by DarkDimius** [comment]
> Maybe everything except the logic that actually creates the synthesized initialize method (i.e., everything except the block inside synthesizeInitialize) and the tests?

that would be perfect!

**Comment by azdavis** [comment]
@jez Bumping this now that #1024 has been merged.

File: dsl/dsl.cc [hunk 2]
```diff
@@ -112,10 +113,41 @@ class DSLReplacer {
...
+                }
+                args.emplace_back(std::move(arg));
+                sigKeys.emplace_back(ast::MK::Symbol(loc, prop.name));
+                sigVals.emplace_back(std::move(prop.type));
+            }
+            classDef->rhs.emplace_back(
+                ast::MK::SigVoid(loc, ast::MK::Hash(loc, std::move(sigKeys), std::move(sigVals))));
+            classDef->rhs.emplace_back(ast::MK::Method(loc, loc, core::Names::initialize(), std::move(args),
+                                                       ast::MK::EmptyTree(), ast::MethodDef::DSLSynthesized));
+        }
```
Comment (jez) [inline dsl/dsl.cc]: This code should almost certainly not be in `dsl.cc`. Can we make "adding `T::Struct#initialize`" feel exactly the same as any other DSL pass?

File: test/testdata/dsl/t_struct_override.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  sig {params(foo: Integer).void}
+  def initialize(foo:)
+    @foo = foo.to_s
+  end
+end
+
+Override.new # error: Missing required keyword argument `foo` for method `Override#initialize`
+Override.new(foo: "no") # error: `String("no")` does not match `Integer` for argument `foo`
+Override.new(foo: 3, bar: 4) # error: Unrecognized keyword argument `bar` passed for method `Override#initialize`
+T.reveal_type(Override.new(foo: 3).foo) # error: Revealed type: `String`
```
Comment (jez) [inline test/testdata/dsl/t_struct_override.rb]: Can you add some `dsl-tree.exp` snapshot tests?

File: dsl/dsl.cc (hunk 2 above)
Comment (azdavis) [reply dsl/dsl.cc]: I see what you mean, and I was a bit uncomfortable with putting this here. But, given that `ChalkODMProp::replaceDSL` only operates on a single call to `prop`, it wasn't clear to me where this code (which I want to run "at the end") should go. Do you have thoughts on how to better structure this?

Comment (jez) [reply dsl/dsl.cc]: The way I did it when I was prototyping this back in October was to use `patchDSL`:

https://git.corp.stripe.com/stripe-internal-archive/sorbet/pull/1597/files#diff-f1d169de1bf307604ed59c9b4b15859eR24

(Also instead of building up something like NodesAndProp, I just factored the prop parsing logic out of the replaceDSL method, so that I could call it from patchDSL.)

Comment (azdavis) [reply dsl/dsl.cc]: I see. I could certainly move the generation of `initialize` to its own `patchDSL` call. However, even if I did that I'd still have to thread the information from the collected vector of `props` back into that `patchDSL` call. Should I try to do something like what you did where you parse the prop information after the fact? (The reason why I initially didn't do that here is because I feel like that would lead to duplication of logic concerning the rules of how the prop DSL works. Perhaps there's a better way to share that information than what I've come up with here?)

Comment (azdavis) [reply dsl/dsl.cc]: Resolved in #1036.

**Comment by azdavis** [comment]
@jez bumping now that the last pay-server pr is in master-passing-tests.

File: test/testdata/dsl/t_struct_default.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# typed: true
+
+class Default < T::Struct
+  prop :foo, Integer, default: 3
+end
+
+T.reveal_type(Default.new.foo) # error: Revealed type: `Integer`
+Default.new(foo: "no") # error: Expected `Integer` but found `String("no")` for argument `foo`
+Default.new(foo: 3, bar: 4) # error: Unrecognized keyword argument `bar` passed for method `Default#initialize`
+T.reveal_type(Default.new(foo: 3).foo) # error: Revealed type: `Integer`
```
Comment (jez) [inline test/testdata/dsl/t_struct_default.rb]: Can you add a test that checks `::T::Struct` as well?

File: dsl/Prop.cc [hunk 5]
```diff
@@ -151,9 +173,16 @@ optional<NodesAndPropInfo> processProp(core::MutableContext ctx, ast::Send *send
...
+    ret.propInfo.name = name;
+    ret.propInfo.type = ASTUtil::dupType(type.get());
+    ret.propInfo.optional = isTNilable(type.get());
 
     // Compute the getters
     if (rules) {
+        if (ASTUtil::hasHashValue(ctx, *rules, core::Names::default_()) ||
+            ASTUtil::hasTruthyHashValue(ctx, *rules, core::Names::factory())) {
+            ret.propInfo.optional = true;
+        }
```
Comment (jez) [inline dsl/Prop.cc]: Can we populate this information somewhere else? This is under the `// Compute the getters` section and more generally the `we're going to be consuming the tree` section. It seems like these operations are read-only.

File: dsl/Prop.cc [hunk 6]
```diff
@@ -325,11 +361,34 @@ void Prop::patchDSL(core::MutableContext ctx, ast::ClassDef *klass) {
...
         ENFORCE(!nodesAndPropInfo->nodes.empty(), "nodesAndPropInfo with value must have nodes be non empty");
         replaceNodes[stat.get()] = std::move(nodesAndPropInfo->nodes);
-        props.emplace_back(nodesAndPropInfo->propInfo);
+        props.emplace_back(std::move(nodesAndPropInfo->propInfo));
     }
     auto oldRHS = std::move(klass->rhs);
     klass->rhs.clear();
     klass->rhs.reserve(oldRHS.size());
+    if (synthesizeInitialize) {
+        // we define ours synthesized initialize first so that if the user wrote one themselves, it overrides ours.
```
Comment (jez) [inline dsl/Prop.cc]:
```suggestion
        // we define our synthesized initialize first so that if the user wrote one themselves, it overrides ours.
```

File: dsl/Prop.cc [hunk 7]
```diff
@@ -325,11 +361,34 @@ void Prop::patchDSL(core::MutableContext ctx, ast::ClassDef *klass) {
...
+        auto loc = klass->loc;
```
Comment (jez) [inline dsl/Prop.cc]: I can imagine this `loc` being super confusing in error messages. Have you tried running sorbet locally on your test cases? Do the type errors look like they could be improved? Like maybe we could use the loc of the respective prop for the loc of each arg.

Comment (azdavis) [reply dsl/Prop.cc]: I like that idea, but what should we do for the loc of the initialize method and its sig? To me the class def seems like the only thing that makes sense.

File: dsl/Prop.cc (hunk 5 above)
Comment (azdavis) [reply dsl/Prop.cc]: Sure. The only reason I put it in here to not introduce another branch on whether `rules` is `nullptr` or not. Should I just not bother trying to optimize a single branch instruction away?

File: test/testdata/dsl/t_struct_default.rb.dsl-tree.exp [hunk 8]
```diff
@@ -0,0 +1,35 @@
+class <emptyTree><<C <root>>> < ()
+  class <emptyTree>::<C Default><<C <todo sym>>> < (<emptyTree>::<C T>::<C Struct>)
+    ::T::Sig::WithoutRuntime.sig() do ||
+      <self>.params({:"foo" => <emptyTree>::<C Integer>}).void()
+    end
+
+    def initialize<<C <todo sym>>>(foo: = ::T.unsafe(nil), &<blk>)
```
Comment (jez) [inline test/testdata/dsl/t_struct_default.rb.dsl-tree.exp]: What do you think about duplicating the default argument into the initialize? That will have the added benefit that this becomes a type error:

```ruby
# typed: true
class A
  prop :foo, Integer, default: ''
end
```

https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20prop%20%3Afoo%2C%20Integer%2C%20default%3A%20''%0Aend

File: test/testdata/dsl/t_struct_override.rb [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  sig {params(foo: Integer).void}
+  def initialize(foo:)
+    @foo = foo.to_s
+  end
+end
+
+Override.new # error: Missing required keyword argument `foo` for method `Override#initialize`
+Override.new(foo: "no") # error: Expected `Integer` but found `String("no")` for argument `foo`
+Override.new(foo: 3, bar: 4) # error: Unrecognized keyword argument `bar` passed for method `Override#initialize`
+T.reveal_type(Override.new(foo: 3).foo) # error: Revealed type: `String`
```
Comment (jez) [inline test/testdata/dsl/t_struct_override.rb]: Do you want to add a test that errors for incompatible override, and assert that the error happens on their `initialize` instead of the class def loc?

File: test/testdata/dsl/t_struct_override.rb [hunk 10]
```diff
@@ -0,0 +1,17 @@
+# typed: true
+
+class Override < T::Struct
+  extend T::Sig
+
+  prop :foo, String
+
+  sig {params(foo: Integer).void}
+  def initialize(foo:)
+    @foo = foo.to_s
```
Comment (jez) [inline test/testdata/dsl/t_struct_override.rb]: This seems like a dangerous test. What do you think about making the body just be `puts 'in overridden initialize'`?

File: test/testdata/dsl/t_struct_some_default.rb [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+class SomeDefault < T::Struct
+  prop :foo, Integer
+  prop :bar, T::Boolean, default: false
+end
+
+SomeDefault.new(foo: 1)
+SomeDefault.new(foo: 2, bar: true)
+SomeDefault.new(foo: 3, bar: false)
+SomeDefault.new # error: Missing required keyword argument `foo` for method `SomeDefault#initialize`
```
Comment (jez) [inline test/testdata/dsl/t_struct_some_default.rb]: These all seem like good simple test cases, but they don't really test edge-casey behaviors. From your experience codemodding pay-server, do you think there's any test you'd like to write to accommodate things that came up? Specifically around inheritance or the things you ended up having to `T.unsafe` or make into an `T::InexactStruct`.

Also, are there test that the non-T::Struct subclasses have untyped initialize methods? If not it would be good to document that behavior too.

File: test/testdata/dsl/t_struct_none.rb.dsl-tree.exp [hunk 12]
```diff
@@ -0,0 +1,17 @@
+class <emptyTree><<C <root>>> < ()
+  class <emptyTree>::<C None><<C <todo sym>>> < (<emptyTree>::<C T>::<C Struct>)
+    ::T::Sig::WithoutRuntime.sig() do ||
+      <self>.params({}).void()
```
Comment (jez) [inline test/testdata/dsl/t_struct_none.rb.dsl-tree.exp]: I'm a little surprised to see this this isn't a type error:

```ruby
class A
  extend T::Sig

  sig {params().void}
  def bar
  end
end
```

^ because this is:

https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Bparams().void%7D%0A%20%20def%20bar%0A%20%20end%0Aend%0A

File: website/docs/tstruct.md [hunk 13]
```diff
@@ -18,17 +18,19 @@ require 'sorbet-runtime'
...
 
 ```ruby
 class MyStruct < T::Struct
-  sig {params(x: Integer, y: String).void}
-  def initialize(x:, y: nil)
+  sig {params(x: Integer, y: String, z: Float).void}
+  def initialize(x:, y: nil, z: 0.5)
     @x = x
     @y = y
+    @z = z
```
Comment (jez) [inline website/docs/tstruct.md]: Thinking about this example, is it even accurate to show people these instance variables?

I don't think that's how the runtime works, and I don't think we want people to be able to know about T::Struct internals even if it does work this way. What do you think about eliding the method bodies of all the DSL-synthesized methods in this page?

File: test/testdata/dsl/t_struct_default.rb.dsl-tree.exp (hunk 8 above)
Comment (azdavis) [reply test/testdata/dsl/t_struct_default.rb.dsl-tree.exp]: Oh nice. I didn't consider that there would be benefits to carting around the default expression but this shows that there are. I'll work on that.

File: test/testdata/dsl/t_struct_override.rb (hunk 10 above)
Comment (azdavis) [reply test/testdata/dsl/t_struct_override.rb]: Sure, but what do you mean by dangerous?

File: test/testdata/dsl/t_struct_default.rb (hunk 4 above)
Comment (azdavis) [reply test/testdata/dsl/t_struct_default.rb]: Added (and fixed a bug).

File: test/testdata/dsl/t_struct_none.rb.dsl-tree.exp (hunk 12 above)
Comment (azdavis) [reply test/testdata/dsl/t_struct_none.rb.dsl-tree.exp]: In this degenerate case, we pass an empty hash to `params` which seems to make it OK. (Trying it on sorbet.run makes it fine there too.)

File: website/docs/tstruct.md (hunk 13 above)
Comment (azdavis) [reply website/docs/tstruct.md]: I agree that it might be good to not expose the inner workings of the runtime machinery by showing the bodies of the generated methods, especially if that's not actually the way they work. But having the bodies might help a user develop a mental model for how the generated methods work (ah, the initialize sets up the internal data the way I expect, and the getter and setter methods get and update the internal data accordingly). It could be argued, however, that even without the method bodies, it's pretty clear what's going on.

Comment (azdavis) [reply website/docs/tstruct.md]: Actually, one case I feel we should perhaps document explicitly is what the setter methods return. Do they return the old value, or the new value? Based on the current example, it looks like they return the new value - is that accurate?

File: test/testdata/dsl/t_struct_some_default.rb (hunk 11 above)
Comment (azdavis) [reply test/testdata/dsl/t_struct_some_default.rb]: Though I did spend a few days going over pay-server to fix bugs surfaced by this change, I'm not sure if I can think of anything very "edge-casey". Most bugs seemed like fairly simple "this argument is not the right type" or had to do with overridden `initialize` which is not a pattern we want to support super well.

I can add a test that asserts that subclasses of `T::InexactStruct` don't get well-typed `new`.

File: test/testdata/dsl/t_struct_override.rb (hunk 9 above)
Comment (azdavis) [reply test/testdata/dsl/t_struct_override.rb]: Sure.

File: website/docs/tstruct.md (hunk 13 above)
Comment (azdavis) [reply website/docs/tstruct.md]: Update: I think I was misunderstanding how Ruby works. It seems that you can't make an `x=` method return anything other than the RHS.

File: test/testdata/dsl/t_struct/inexact.rb [hunk 14]
```diff
@@ -0,0 +1,9 @@
+# typed: true
+
+class Inexact < T::InexactStruct
+  prop :foo, Integer
+  prop :bar, String
+end
+
+Inexact.new
+Inexact.new(foo: 3, bar: "hey") # error: Wrong number of arguments for constructor. Expected: `0`, got: `1`
```
Comment (jez) [inline test/testdata/dsl/t_struct/inexact.rb]: Why is this an error? I feel like this is a regression if this is an error.

Comment (jez) [reply test/testdata/dsl/t_struct/inexact.rb]: Are we missing something in our RBI in sorbet?

Comment (azdavis) [reply test/testdata/dsl/t_struct/inexact.rb]: Resolved offline: This has to do with the fact that we don't use sorbet-runtime in pay-server, so we never actually noticed this error was being emitted (which it shouldn't be.)
